docker/go-control-plane/fallback-control-plane.go

```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	return snapshot, nil
+}
+
+// Abruptly stops the server when the client requests a resource that the test
+// marked as one that should trigger this behavior
+func (cb *controlService) onStreamRequest(id int64, req *v3discoverypb.DiscoveryRequest) error {
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	log.Printf("Received request for %s on stream %d: %v:%v\n", req.GetTypeUrl(), id, req.VersionInfo, req.ResourceNames)
+	filtered := cb.filters[req.GetTypeUrl()]
+	if filtered != nil {
+		for _, name := range req.ResourceNames {
+			if filtered[name] {
+				log.Printf("Self destructing: %s/%s\n", req.GetTypeUrl(), name)
+				os.Exit(0)
+			}
+		}
+	}
+	return nil
+}
+
+func RunServer(srv server.Server, controlService xdsconfigpb.XdsConfigControlServiceServer, port uint) error {
+	grpcServer := grpc.NewServer()
+	reflection.Register(grpcServer)
+	channelz.RegisterChannelzServiceToServer(grpcServer)
+	xdsconfigpb.RegisterXdsConfigControlServiceServer(grpcServer, controlService)
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return err
+	}
+	v3discoverypb.RegisterAggregatedDiscoveryServiceServer(grpcServer, srv)
+	log.Printf("Management server listening on %d\n", port)
+	if err = grpcServer.Serve(lis); err != nil {
+		log.Println(err)
+	}
+	return nil
+}
+
+// Main entry point. Configures and starts a gRPC server that serves xDS traffic
+// and provides an interface for tests to manage control plane behavior.
+func main() {
+	flag.Parse()
+	host, upstreamPort, err := net.SplitHostPort(*upstream)
+	if err != nil {
+		log.Fatalf("Incorrect upstream host name: %+v: %+v\n", upstream, err)
+	}
+	parsedUpstreamPort, err := strconv.Atoi(upstreamPort)
+	if err != nil || parsedUpstreamPort <= 0 {
+		log.Fatalf("Not a valid port number: %+v: %+v\n", upstreamPort, err)
+	}
+	// The type needs to be checked
+	controlService := &controlService{version: 1,
+		clusters:  map[string]*v3clusterpb.Cluster{controlplane.ListenerName: controlplane.MakeCluster(controlplane.ClusterName, host, uint32(parsedUpstreamPort))},
+		listeners: map[string]*v3listenerpb.Listener{controlplane.ListenerName: controlplane.MakeHTTPListener(controlplane.ListenerName, controlplane.ClusterName)},
+		filters:   map[string]map[string]bool{},
+		cache:     cache.NewSnapshotCache(false, cache.IDHash{}, nil),
+	}
+	// Create a cache
+	snapshot, err := controlService.MakeSnapshot()
+	if err != nil {
+		log.Fatalf("Snapshot error %q for %+v\n", err, snapshot)
+	}
+	// Add the snapshot to the cache
+	if err := controlService.cache.SetSnapshot(context.Background(), *nodeid, snapshot); err != nil {
+		log.Fatalf("Snapshot error %q for %+v\n", err, snapshot)
+	}
+
+	// Run the xDS server
+	ctx := context.Background()
+	srv := server.NewServer(ctx, controlService.cache, server.CallbackFuncs{
+		StreamRequestFunc: controlService.onStreamRequest,
+	})
+	err = RunServer(srv, controlService, *port)
+	if err != nil {
+		log.Fatalf("Server startup failed: %q\n", err)
+	}
```

Comment:
This could be on a single line:

```golang
	if err := RunServer(srv, controlService, *port); err != nil {
		log.Fatalf("Server startup failed: %q\n", err)
	}
```